File: day-4/main.py
with open("input.txt") as f:
    data = f.read().strip().split()

fullyContains = 0
rangesOverlaps = 0

for line in data:
    elves = line.split(",")
    ranges = [list(map(int, elf.split("-"))) for elf in elves]
    print(ranges)

    elf1Start, elf1End = ranges[0]
    elf2Start, elf2End = ranges[1]

    cond1 = (elf1Start <= elf2Start) and (elf1End >= elf2End)
    cond2 = (elf1Start >= elf2Start) and (elf1End <= elf2End)

    if cond1 or cond2:
        fullyContains += 1
    
    cond3 = elf1Start > elf2End
    cond4 = elf1End < elf2Start

    if not (cond3 or cond4):
        rangesOverlaps += 1

print("Fully contains:", fullyContains)
print("Ranges overlaps:", rangesOverlaps)

File: day-4/main.go
package main

import (
	"bufio"
	"fmt"
	"os"
)

func isContained(a, b, c, d int) bool {
	if a >= c && b <= d {
		return true
	}

	return false
}

func overlaps(a, b, c, d int) bool {
	if !(a > d || b < c) {
		return true
	}

	return false
}

func main() {
	file, err := os.Open("input.txt")

	if err != nil {
		panic(err)
	}

	defer file.Close()

	fullyContained, rangeOverlaps := 0, 0
	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		elf1Start, elf1End, elf2Start, elf2End := 0, 0, 0, 0
		_, err := fmt.Sscanf(scanner.Text(), "%d-%d,%d-%d", &elf1Start, &elf1End, &elf2Start, &elf2End)
		if err != nil {
			panic(err)
		}

		if isContained(elf1Start, elf1End, elf2Start, elf2End) || isContained(elf2Start, elf2End, elf1Start, elf1End) {
			fullyContained++
		}

		if overlaps(elf1Start, elf1End, elf2Start, elf2End) {
			rangeOverlaps++
		}

	}
	fmt.Println("Fully Contains:", fullyContained)
	fmt.Println("Range Overlaps:", rangeOverlaps)

}
